### PR TITLE
8321387: SegmentAllocator:allocateFrom(AddressLayout, MemorySegment) does not throw stated UnsupportedOperationException

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to change the exception type for exceptions thrown for certain methods with a parameter of type `MemorySegment` when it is `MemorySegment::isReadOnly`. Previously an `UnsupportedOperationException` was specified but in some cases, in reality, an `IllegalArgumentException` was thrown. 

The principle used in this PR is that operations acting on an MS where the MS is `this` should throw an `UnsupportedOperationException` whereas in cases where the MS is a *parameter* an `IllegalArgumentException` should be thrown.

It should be noted that this PR retains the previous behavior for MS VarHandle access (even though the MS is a parameter to the accessor methods, the first parameter can be said to represent `this`).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8321513](https://bugs.openjdk.org/browse/JDK-8321513) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8321387](https://bugs.openjdk.org/browse/JDK-8321387): SegmentAllocator:allocateFrom(AddressLayout, MemorySegment) does not throw stated UnsupportedOperationException (**Bug** - P3)(⚠️ The fixVersion in this issue is [22] but the fixVersion in .jcheck/conf is 23, a new backport will be created when this pr is integrated.)
 * [JDK-8321513](https://bugs.openjdk.org/browse/JDK-8321513): SegmentAllocator:allocateFrom(AddressLayout, MemorySegment) does not throw stated UnsupportedOperationException (**CSR**)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16993/head:pull/16993` \
`$ git checkout pull/16993`

Update a local copy of the PR: \
`$ git checkout pull/16993` \
`$ git pull https://git.openjdk.org/jdk.git pull/16993/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16993`

View PR using the GUI difftool: \
`$ git pr show -t 16993`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16993.diff">https://git.openjdk.org/jdk/pull/16993.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16993#issuecomment-1842942975)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java [hunk 1]
```diff
@@ -358,10 +359,19 @@ private <Z> Z toArray(Class<Z> arrayClass, ValueLayout elemLayout, IntFunction<Z
         return arr;
     }
 
+    public enum AccessConstraint {
+        READ_ONLY, ON_WRITE_UOE, ON_WRITE_IAE
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java:363]: I think the names are a bit confusing. Perhaps having NONE instead of READ_ONLY would be better.

File: src/java.base/share/classes/java/lang/invoke/X-VarHandleSegmentView.java.template [hunk 2]
```diff
@@ -149,7 +150,7 @@ final class VarHandleSegmentAs$Type$s extends VarHandleSegmentViewBase {
     @ForceInline
     static void set(VarHandle ob, Object obb, long base, $type$ value) {
         VarHandleSegmentViewBase handle = (VarHandleSegmentViewBase)ob;
-        AbstractMemorySegmentImpl bb = checkAddress(obb, base, handle.length, false);
+        AbstractMemorySegmentImpl bb = checkAddress(obb, base, handle.length, ON_WRITE_UOE);
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/invoke/X-VarHandleSegmentView.java.template:153]: This is actually tricky: for var handles it seems like the right exception is IAE - but then for MemorySegment::get the right exception is UOE (because the receiver itself is read-only). We might need to give up consistency a bit, as otherwise I don't think memory segment getter/setter can reuse the VarHandle implementation.

[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/invoke/X-VarHandleSegmentView.java.template:153]: After offline discussion: the distinction between UOE and IAE is mostly fictional and doesn't add much. In fact, such distinction is harmful because it would mean that code written using var handles cannot cleanly migrate to use memory segment getter/setter (and vice-versa). As such, I recommend that the implementation is left alone, and IAE is thrown by the var handle accessors for read-only segments. Memory segment setters should similarly throw IAE for read-only, because the setters are just thin easy to use wrappers around the var handle.

A similar argument holds for MemorySegment::copyFrom - here throwing UOE would be more apt, but given we explain this method as calling the static MemorySegment::copy (which throws IAE) I would again avoid exception mismatch which would be detrimental to refactoring.

At which point we're left with MemorySegment::fill. While UOE would be better here, it would be the only method in MS issuing UOE for a read-only issue, so I think it would be just best to regularize and use IAE here too.

tl;dr; the implementation is already throwing the correct set of exceptions (which is not perfect, but better than the alternatives). The javadoc should be updated to reflect that.

File: src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java [hunk 3]
```diff
@@ -92,6 +92,8 @@ public interface SegmentAllocator {
      *
      * @param str the Java string to be converted into a C string
      * @return a new native segment containing the converted C string
+     * @throws IllegalArgumentException if the allocated segment is
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/SegmentAllocator.java:95]: I don't think the changes here are useful. What does it mean for an allocated segment to be read-only? I think all these conditions are tied to `prefixAllocator` blindly accepting read-only segments, which should NOT be the case. I suggest to revert all the chnages here and document (and throw) a new exception for when a prefix allocator is created from a read-only segment.

File: src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java [hunk 4]
```diff
@@ -358,10 +359,19 @@ private <Z> Z toArray(Class<Z> arrayClass, ValueLayout elemLayout, IntFunction<Z
...
 
+    public enum AccessConstraint {
+        NONE, ON_WRITE_UOE, ON_WRITE_IAE
+    }
+
     @ForceInline
-    public void checkAccess(long offset, long length, boolean readOnly) {
-        if (!readOnly && this.readOnly) {
-            throw new UnsupportedOperationException("Attempt to write a read-only segment");
+    public void checkAccess(long offset, long length, AccessConstraint access) {
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java:367]: These changes should be reverted, I don't think we use the UOE mode anymore. Just revert the impl to what it was before (also true for VarHandle templates).

**Review by mcimadamore — COMMENTED** [review]
There are still more changes than required, I think

**Review by mcimadamore — APPROVED** [review]
Looks good!
